### PR TITLE
chore(core): introduce internal off-heap data structures

### PR DESCRIPTION
**Pull request by puzpuzpuz**
Introduces `DirectIntList`

### CODE REVIEW

File: core/src/main/java/io/questdb/std/DirectIntList.java [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+    }
+
+    public void add(int x) {
+        ensureCapacity();
+        assert pos < limit;
+        Unsafe.getUnsafe().putInt(pos, x);
+        pos += Integer.BYTES;
+    }
+
+    public final void add(DirectIntList that) {
```
Comment (jerrinot) [inline core/src/main/java/io/questdb/std/DirectIntList.java:61]: nit: `allAll()` as in `IntList` or even `j.u.List`

File: core/src/main/java/io/questdb/std/DirectIntList.java [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+    public void setCapacity(long capacity) {
+        setCapacityBytes(capacity * Integer.BYTES);
+    }
+
+    public void setPos(long p) {
+        assert p * Integer.BYTES <= capacity;
+        pos = start + p * Integer.BYTES;
+    }
+
+    public void shrink(long newCapacity) {
```
Comment (jerrinot) [inline core/src/main/java/io/questdb/std/DirectIntList.java:132]: not tested

Comment (jerrinot) [reply core/src/main/java/io/questdb/std/DirectIntList.java:132]: also it's unclear whether the capacity is in bytes or in number of int slots

File: core/src/main/java/io/questdb/std/DirectIntList.java [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+    }
+
+    public long size() {
+        return (int) ((pos - start) / Integer.BYTES);
+    }
+
+    @Override
+    public String toString() {
+        CharSink sb = Misc.getThreadLocalBuilder();
+        sb.put('{');
```
Comment (jerrinot) [inline core/src/main/java/io/questdb/std/DirectIntList.java:146]: not: the `io.questdb.std.IntList` uses `[`, not `{`. 
consider making this consistent.

File: core/src/main/java/io/questdb/std/DirectIntList.java [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+    public void resetCapacity() {
+        setCapacityBytes(initialCapacity);
+    }
+
+    public void set(long p, int v) {
+        assert p >= 0 && p <= (limit - start) >> 2;
+        Unsafe.getUnsafe().putInt(start + (p << 2), v);
+    }
+
+    // desired capacity in LONGs (not count of bytes)
```
Comment (jerrinot) [inline core/src/main/java/io/questdb/std/DirectIntList.java:122]: should be INTs, not LONGs. copy'n'paste error?

File: core/src/main/java/io/questdb/std/DirectIntList.java [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+        ensureCapacity();
+        assert pos < limit;
+        Unsafe.getUnsafe().putInt(pos, x);
+        pos += Integer.BYTES;
+    }
+
+    public final void add(DirectIntList that) {
+        long thatSize = that.pos - that.start;
+        if (limit - pos < thatSize) {
+            setCapacityBytes(this.capacity + thatSize - (limit - pos));
```
Comment (jerrinot) [inline core/src/main/java/io/questdb/std/DirectIntList.java:64]: this sizes the capacity precisely as needed to accommodate for `that` list.
this means the next `add()` is guaranteed to trigger another reallocation. Is this intentional?

File: core/src/main/java/io/questdb/std/DirectIntList.java [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+        setCapacityBytes(initialCapacity);
+    }
+
+    public void set(long p, int v) {
+        assert p >= 0 && p <= (limit - start) >> 2;
+        Unsafe.getUnsafe().putInt(start + (p << 2), v);
+    }
+
+    // desired capacity in LONGs (not count of bytes)
+    public void setCapacity(long capacity) {
```
Comment (jerrinot) [inline core/src/main/java/io/questdb/std/DirectIntList.java:123]: nit: when you set `capacity` to `0` then the list becomes unusable - as `ensureCapacity()` resizes it to `0 * 2` :)

**Review by jerrinot — CHANGES_REQUESTED** [review]
mostly trivial changes + some missing test coverage. 
the core impl is sound.

File: core/src/main/java/io/questdb/std/DirectIntList.java (hunk 5 above)
Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/std/DirectIntList.java:64]: Yes, I think so. In many cases when we copy an off-heap list, we simply need the exact copy and we don't expand it after that.

File: core/src/main/java/io/questdb/std/DirectIntList.java (hunk 1 above)
Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/std/DirectIntList.java]: Renamed in 4f6970662d42b81a7edd80f348c73eb4ed238e34

File: core/src/main/java/io/questdb/std/DirectIntList.java (hunk 2 above)
Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/std/DirectIntList.java:131]: Fixed a bug and added a test in 4f6970662d42b81a7edd80f348c73eb4ed238e34

File: core/src/main/java/io/questdb/std/DirectIntList.java (hunk 3 above)
Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/std/DirectIntList.java]: Changed in 4f6970662d42b81a7edd80f348c73eb4ed238e34

File: core/src/main/java/io/questdb/std/DirectIntList.java [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+import io.questdb.std.str.CharSink;
+
+import java.io.Closeable;
+
+public class DirectIntList implements Mutable, Closeable, Reopenable {
+
+    private static final Log LOG = LogFactory.getLog(DirectIntList.class);
+    private final long initialCapacity;
+    private final int memoryTag;
+    private long address;
```
[the remark this answers is not in the document]
Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/std/DirectIntList.java:39]: That looks like a leftover. Removed in 4f6970662d42b81a7edd80f348c73eb4ed238e34

File: core/src/main/java/io/questdb/std/DirectIntList.java [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+        if (address == 0) {
+            resetCapacity();
+        }
+    }
+
+    public void resetCapacity() {
+        setCapacityBytes(initialCapacity);
+    }
+
+    public void set(long p, int v) {
```
[the remark this answers is not in the document]
Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/std/DirectIntList.java:115]: Added a test in 4f6970662d42b81a7edd80f348c73eb4ed238e34

File: core/src/main/java/io/questdb/std/DirectIntList.java (hunk 4 above)
Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/std/DirectIntList.java]: Yes, that was a copy-paste leftover. Fixed in 4f6970662d42b81a7edd80f348c73eb4ed238e34

File: core/src/main/java/io/questdb/std/DirectIntList.java (hunk 6 above)
Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/std/DirectIntList.java:121]: Added an assertion in 4f6970662d42b81a7edd80f348c73eb4ed238e34

**Comment by puzpuzpuz** [comment]
All addressed suggestions were also applied to `DirectLongList` in 4f6970662d42b81a7edd80f348c73eb4ed238e34

**Review by jerrinot — APPROVED** [review]

**Comment by puzpuzpuz** [comment]
@jerrinot could you take another look and reapprove? I fixed a few broken tests in 96bed3a4bb969d163826b4c4c77452ba12084ef4

**Review by jerrinot — APPROVED** [review]

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 93 / 98 (94.90%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/DirectIntList.java|75|80|93.75%|
|:large_blue_circle:|io/questdb/std/DirectLongList.java|18|18|100.00%|

**Pull request merged**